scipy/optimize/_gensa.py

```diff
@@ -0,0 +1,645 @@
+                # We get a better energy value
+                self._etot0 = np.array(self._etot)
+                if self._etot < self._emini:
+                    self._emini = np.array(self._etot)
+                    self._xmini = np.array(self._x)
+                    self._emini_unchanged = False
+                    self._index_no_emini_update = 0
+            else:
+                # We do not have improvement but do we accept location
+                r = self._random_state.random_sample()
+                pqa1 = (self.qa - 1.0) * (
+                    self._etot - self._etot0) / self._temp_qa + 1.0
+                if pqa1 < 0.0:
+                    pqa = 0.0
+                else:
+                    pqa = np.exp(np.log(pqa1) / (1.0 - self.qa))
+                if r > pqa:
+                    # We reject the new visiting location
+                    self._x = self._xbackup
+                else:
+                    # The new visiting location is accepted
+                    self._etot0 = self._etot
+            if self._check_stopping():
+                self._stop_search()
+                return 0
+            if self._index_no_emini_update >= self._index_tol_emini_update - 1:
+                if j == 0:
+                    self._emini_markov = np.array(self._etot0)
+                    self._xmini_markov = np.array(self._x)
+                else:
+                    if self._etot0 < self._emini_markov:
+                        self._emini_markov = np.array(self._etot0)
+                        self._xmini_markov = np.array(self._x)
+
+    def start_search(self):
+        """ Start annealing process with eventually re-annealing
+        """
+        in_constraint = True
+        self._emini_unchanged = True
+        self._emini_markov = 0.0
+        self._xmini_markov = np.zeros(self._x.size)
+        self._index_no_emini_update = 0
+        self._index_tol_emini_update = 1000
+        self._starttime = time.time()
+        self._emini = np.array(self._etot)
+        self._xmini = np.array(self._x)
+        self._etot0 = np.array(self._etot)
+        if self._etot < self._emini:
+            self._emini = np.array(self._etot)
+            self._xmini = np.array(self._x)
+        self._etot0 = self._etot
+        if self._check_stopping():
+            self._stop_search()
+        self._step_record = 0
+        self._temp = self.temp_start
+        tolowtemp = True
+        while(tolowtemp):
+            for i in range(self.maxsteps):
+                # Evaluating iteration artificial temperature
+                s = float(i) + 2.0
+                t1 = np.exp((self.qv - 1) * np.log(2.0)) - 1.0
+                t2 = np.exp((self.qv - 1) * np.log(s)) - 1.0
+                self._temp = self.temp_start * t1 / t2
+                self._step_record += 1
+                self._temp_qa = self._temp / float(i + 1)
+                self._index_no_emini_update += 1
+                # Maximum number of iterations reached
```

Comment:
maybe extend the comment a bit: "break out of both for-loop and while loop". and then a few lines lower "break out of for-loop only".